### ISSUE TITLE
Road to PKI/signing/certifying pdf

### ISSUE THREAD

**Issue by colourful-land** [qpdf/qpdf#201]
I know it's a lot of features and lots of work to get these done, but maybe we can

1. take advantage of the recent release ISO 32000 PDF2.0 to seek some shortcut, prioritize the support on the future pdf standard v.s. backwards compatible details.
2. break down the features into several smaller ones and choose to take the low hanging fruit.
3. do some fundraising about it.

There is competition even among opensource software, and no others have provided signing or verifying capacity for cryptographically signed PDF. If qpdf can do it, a lot of opensource software (e.g. the default pdf viewer evince of Gnome) will depend on it, making it available on every platform.

**Comment by michael-o** [comment]
Overview for signatures: https://www.adobe.com/devnet-docs/etk_deprecated/tools/DigSig/Acrobat_DigitalSignatures_in_PDF.pdf

**Comment by jberkenbilt** [comment]
@michael-o Thanks for posting this.

**Comment by michael-o** [comment]
I have grabbed now a copy of ISO 32000 and will try to understand in the next couple of weeks whether signatures always required forms and fields.

**Comment by colourful-land** [comment]
Hi let me know if someone is able to tackle this problem, I'm interested in sponsoring.

**Comment by jberkenbilt** [comment]
I definitely don't want to discourage contributions, but I am pretty tight about accepting contributions. Since I am the sole developer/maintainer on qpdf, any contribution I accept becomes my problem to maintain. I have accepted a few significant contributions over the years. I would like to work on this feature. My guess is that it wouldn't take me very long. It's good to know that poppler can do it now. While I don't look at other implementations' source code (mainly for licensing reasons), it's good to have a variety of tools for validation purposes. Anyway qpdf and poppler aren't really "competitors" since poppler has much more functionality (e.g. ability to fully render PDF, do text extraction, etc.) and also has a more restrictive license (GPL vs. Apache for qpdf). However I definitely think it's important for qpdf to get this functionality.

**Comment by timolehto** [comment]
Any news on this? And, since I'm not really interested in the sining so much as I'm interested in decrypting, do you have any idea how long way would we be from PKI decrpytion/encryption support if we would have support for certificates and signing?

**Comment by jberkenbilt** [comment]
I'd say this is second on my list now after the pages work. I may write a blog post about it. I have a proof of concept of verifying the kind of signatures that Adobe's eSignature uses. I haven't attempted to do the new kinds that PDF 2.0 supports. My intention is to write a blog post that shows how you can verify digital signatures using qpdf, standard Linux (or mac) shell commands, and openssl. If my schedule allows me to put time into this, signature verification could come in the not-too-distant future, and maybe some limited form of signing could follow. I wish I could give you a time frame, but I can't. Adding significant features to qpdf is something I have to fit in around the many other things I do.

**Comment by jberkenbilt** [comment]
@byF Yes, it will just take a weekend when I have a few hours to write it up. :-) I can bump that up to the front of the list. I had one on the list in front of it, but knowing someone would actually like to see the one on this, I moved it to the front. I made a note to comment here when I write the post, or you can find my blog at https://medium.com/@jberkenbilt.
